### PR TITLE
Add support for the "hashPath" property in the .deps.json file

### PR DESCRIPTION
**Pull request by joelverhagen**
Similar to my previous PR (https://github.com/dotnet/core-setup/pull/235), this adds the `"hashPath"` property to the .deps.json file. This allows the CLI to pass down the path to the hash file instead of [DotNetHost building it itself](https://github.com/dotnet/core-setup/blob/989f0ec1319529b337caf93a6a333bec1551a088/src/corehost/cli/deps_entry.cpp#L168-L175).

I have raised this issue to @gkhanna79 and folks over email and we have not yet come to a conclusion on this course correction, but I wanted to get this change ready just in case.

**Edit:** I have also make the .deps.json parser allow unknown extra properties in objects. This was behavior was supported in version 1.0.0 of DependencyModel, but regressed after some performance work (moving from `JObject` to `JsonTextReader`).

/cc @gkhanna79 @livarcocc @eerhardt @pakrym

### CODE REVIEW

**Comment by eerhardt** [comment]
I think this looks good.

One question:  was there a critical reason why we needed to "ToLower" the hash file name?

**Comment by joelverhagen** [comment]
The hash file is used by NuGet as a marker indicating that the package is fully extracted and therefore needs to be at a well-known (case-insensitive) location.

File: src/Microsoft.Extensions.DependencyModel/JsonTextReaderExtensions.cs [hunk 1]
```diff
@@ -16,7 +16,19 @@ internal static bool TryReadStringProperty(this JsonTextReader reader, out strin
...
+                
+                if (reader.Read())
+                {
+                    if (reader.TokenType == JsonToken.String)
+                    {
+                        value = (string)reader.Value;
+                    }
+                    else
+                    {
+                        reader.Skip();
```
Comment (livarcocc) [inline src/Microsoft.Extensions.DependencyModel/JsonTextReaderExtensions.cs]: Should this return true even when we skip reading the string value?

**Comment by livarcocc** [comment]
:shipit:

File: src/Microsoft.Extensions.DependencyModel/JsonTextReaderExtensions.cs (hunk 1 above)
Comment (joelverhagen) [inline src/Microsoft.Extensions.DependencyModel/JsonTextReaderExtensions.cs]: Yes, the `true` indicates that the caller should keep probing for more properties.

**Comment by pakrym** [comment]
:shipit:

**Pull request merged**